cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,1164 @@
+      // avoid extra panalization. One way to achieve it is by calling thrust::set_intersection
+      // to filter out the removed edges. However this will require another array.
+      
+      // Find the intersection of 'invalid_edges_buffer' and 'edges' to extract the remaining invalid
+      // edges that still need to be processed. Didn't used thrust::set_intersection because I didn't
+      // want to create a temporary array
+      auto invalid_edge_last =
+        thrust::partition(handle.get_thrust_policy(),
+                          get_dataframe_buffer_begin(invalid_edges_buffer),
+                          get_dataframe_buffer_end(invalid_edges_buffer),
+                          [edge_first = edges, // rename to 'edges'
+                           edge_last = edges + num_edges,
+                           num_edges = num_edges]
+                           __device__(auto invalid_edge) {
+                            auto itr = thrust::find(thrust::seq, edge_first, edge_last, invalid_edge);
+                            auto idx = thrust::distance(edge_first, itr);
+                            return idx < num_edges;
+                          });
+      
+      num_invalid_edges = thrust::distance(get_dataframe_buffer_begin(invalid_edges_buffer), invalid_edge_last);
+
+
+      resize_dataframe_buffer(
+        invalid_edges_buffer, num_edges, handle.get_stream());
+
+      // Need to run prefix_sum again to get new ranges because some incoming edges were removed
+      prefix_sum.resize(num_invalid_edges + 1, handle.get_stream());
+      indices.resize(num_invalid_edges, handle.get_stream());
+
+      // FIXME: need to sort 'incoming_vertex_pairs'. No need because a stable partition was
+      // performed that preserve the sorting
+      thrust::tabulate(
+        handle.get_thrust_policy(),
+        prefix_sum.begin(),
+        prefix_sum.end(),
+        [invalid_first               = get_dataframe_buffer_begin(invalid_edges_buffer),
+         dst_array_begin = std::get<0>(incoming_vertex_pairs).begin(),
+         num_edges                   = num_edges] __device__(auto idx) {
+          auto src = thrust::get<0>(*(invalid_first + idx));
+          auto dst = thrust::get<1>(*(invalid_first + idx));
+          auto dst_array_end = dst_array_begin + num_edges;
+          auto itr_lower = thrust::lower_bound(thrust::seq, dst_array_begin, dst_array_end, dst);
+          auto idx_lower =
+            thrust::distance(dst_array_begin, itr_lower);  // FIXME: remove self loops
+          auto itr_upper = thrust::upper_bound(thrust::seq, dst_array_begin, dst_array_end, dst);
+          auto idx_upper = thrust::distance(dst_array_begin, itr_upper);
+          auto dist      = thrust::distance(itr_lower, itr_upper);
+          return dist;
+        });
+
+      thrust::exclusive_scan(
+        handle.get_thrust_policy(), prefix_sum.begin(), prefix_sum.end(), prefix_sum.begin());
+      
+      // case 3 unroll (p, r)
```

Comment:
I think the code for case 2 & case 3 have a lot in common, can you refactor to reduce the code size?